docs2/site/docs/getting-started/custom-scalars.md

```diff
@@ -183,105 +213,303 @@ public class ExampleSchema : Schema
+2. Change `Serialize` to return structured data.
 
 ```csharp
-// In ExampleSchema
-public ExampleSchema(IServiceProvider provider)
-    : base(provider)
-{        
-    ValueConverter.Register<string, Vector3>(ParseVector3);
-    RegisterValueConverter(new Vector3AstValueConverter());
-
-    // Other schema assignments...
+// In Vector3Type
+
+public override object Serialize(object value)
+{
+    if (value == null)
+        return null;
+
+    if (value is Vector3 vector3)
+        return vector3;
+
+    return ThrowSerializationError(value);
 }
-``` 
+```
+
+3. Change `ToAST` to return an AST literal that represents the data.
 
-Update the implementation of `Vector3Type.ParseLiteral`:
+Since `Serialize` no longer returns a type that can be converted to an AST node, it is
+necessary to override this method.
 
 ```csharp
 // In Vector3Type
-public override object ParseLiteral(IValue value)
+
+public override IValue ToAST(object value)
 {
-    // new test
-    if (value is Vector3Value vector3Value)
-        return ParseValue(vector3Value.Value);
+    if (value == null)
+        return new NullValue();
+
+    if (value is Vector3 vector3)
+    {
+        return new StringValue($"{vector3.X},{vector3.Y},{vector3.Z}");
+    }
 
-    return value is StringValue stringValue
-        ? ParseValue(stringValue.Value)
-        : null;
+    return ThrowASTConversionError(value);
 }
+```
+
+With these changes, a variable can be parsed as a string or as a structured object, and is always returned as a structured object.
+
+## ValueConverter
+
+When `GetArgument<T>` is called, the argument value is coerced to the requested type via the `ValueConverter`.
+No conversion takes place when the requested type matches the type of the object or scalar (the type returned from
+`ParseLiteral` or `ParseValue`). But you can also use the value converter to assist with input deserialization.
+
+For instance, you may be using `IdGraphType` within your schema as unique identifiers for your data objects. Pursuant
+to the GraphQL specification, these identifiers should be passed as strings such as in the below example:
 
+```graphql
+{
+    widget (id: "3") {
+        name
+    }
+}
 ```
 
-This is necessary since the query executor converts all arguments and variables to `IValue` instances before coercing them to their server-side representation using `ParseLiteral`. `Vector3` instances parsed from variables will be converted to the more specific `Vector3Value` type.
+However, your code may use integer identifiers in the data access layer. So when you call `context.GetArgument<int>("id")`,
+GraphQL.NET calls the value converter to convert the string to an integer.
 
-7. Implement `ScalarGraphType.Serialize` so `Vector3` instances can be sent to the client.
+The value converter can be extended globally by calling the static method `Register` as follows:
```

Comment:
This API is one to think about changing from static to instance method (v5).

Reply:
Or delete it altogether...

Reply:
Resolvefieldcontext.GetArgument still should be rather agile.